numpyro/infer/hmc_gibbs.py

```diff
@@ -476,48 +503,290 @@ class HMCECS(HMCGibbs):
+                                 potential_energy=pe),
+        hmc_state = cond(transition,  # only update gradient and potential energy when block is updated
+                         state.hmc_state._replace(
+                             z_grad=self.grad_fn(partial(potential_fn, z_gibbs, gibbs_state))(state.hmc_state.z),
+                             potential_energy=pe), identity,
+                         state.hmc_state, identity
+                         )
+
+        model_kwargs["_gibbs_sites"] = z_gibbs
+        model_kwargs["_gibbs_state"] = gibbs_state
+        hmc_state = self.inner_kernel.sample(hmc_state, model_args, model_kwargs)
 
         z = {**z_gibbs, **hmc_state.z}
+        return HMCECSState(z, hmc_state, rng_key, gibbs_state, accept_prob)
 
-        return HMCGibbsState(z, hmc_state, rng_key)
+    @property
+    def grad_fn(self):
+        if not hasattr(self, '_grad_mode'):
+            if self.inner_kernel._forward_mode_differentiation:
+                grad_fn = jacfwd
+            else:
+                grad_fn = grad
+            self._grad_fn = grad_fn
+        return self._grad_fn
+
+
+def perturbed_method(subsample_plate_sizes, proxy_fn):
+    def estimator(likelihoods, params, gibbs_state):
+        subsample_log_liks = defaultdict(float)
+        for (fn, value, name, subsample_dim) in likelihoods.values():
+            subsample_log_liks[name] += _sum_all_except_at_dim(fn.log_prob(value), subsample_dim)
+
+        log_lik_sum = 0.
+
+        proxy_value_all, proxy_value_subsample = proxy_fn(params, subsample_log_liks.keys(), gibbs_state)
+
+        for name, subsample_log_lik in subsample_log_liks.items():  # loop over all subsample sites
+            n, m = subsample_plate_sizes[name]
+
+            diff = subsample_log_lik - proxy_value_subsample[name]
+
+            unbiased_log_lik = proxy_value_all[name] + n * jnp.mean(diff)
+            variance = n ** 2 / m * jnp.var(diff)
+            log_lik_sum += unbiased_log_lik - 0.5 * variance
+        return log_lik_sum
+
+    return estimator
+
+
+def taylor_proxy(reference_params):
+    def construct_proxy_fn(prototype_trace, subsample_plate_sizes, model, model_args, model_kwargs, num_blocks=1):
+        ref_params = {name: _unconstrain_reparam(reference_params, site) for name, site in prototype_trace.items()
+                      if name in reference_params}
+        ref_params_flat, unravel_fn = ravel_pytree(ref_params)
+
+        def log_likelihood(params_flat, subsample_indices=None):
+            if subsample_indices is None:
+                subsample_indices = {k: jnp.arange(v[0]) for k, v in subsample_plate_sizes.items()}
+            params = unravel_fn(params_flat)
+            with warnings.catch_warnings():
+                warnings.simplefilter("ignore")
+                with block(), trace() as tr, substitute(data=subsample_indices), substitute(data=params):
```

Comment:
Do I understand correctly that `params_flat` is unconstrained params, which are params that we want to sample from MCMC? If so, we should convert params to constrained values. One way is to use `_unconstrained_reparams`, but you can also use
```
params = {name: biject_to(prototype_trace[name]["fn"].support)(value)
          for name, value in params.items()}
```
to save some unnecessary `log_abs_det_jacobian` terms in `_unconstrained_reparam`.

Reply:
Make sense, I'll revert it.